Add article about running commands in the shell (#9154)
MicrosoftDocs/PowerShell-Docs: reference/docs-conceptual/learn/shell/running-commands.md

## Types of commands

For any shell in any operating system there are three types of commands:

- **Shell language keywords** are part of the shell's scripting language.

  - Examples of `bash` keywords include: `if`, `then`, `else`, `elif`, and `fi`.
  - Examples of `cmd.exe` keywords include: `dir`, `copy`, `move`, `if`, and `echo`.
  - Examples of PowerShell keywords include: `for`, `foreach`, `try`, `catch`, and `trap`.

  Shell language keywords can only be used within the runtime environment of the shell. There is no
  executable file, external to the shell, that provides the keyword's functionality.

- **OS-native commands** are executable files installed in the operating system. The executables can
  be run from any command-line shell, like PowerShell. This includes script files that may require
  other shells to work properly. For example, if you run a Windows batch script (`.cmd` file) in
  PowerShell, PowerShell runs `cmd.exe` and passes in the batch file for execution.

- **Shell environment-specifc commands** are commands defined in external files that can only be
  used within the runtime environment of the shell. These include scripts and functions, or they can

>   be specially compiled modules that add commands to the shell runtime. In PowerShell, these command

Reviewer: ```suggestion
  be specially compiled modules that add commands to the shell runtime. In PowerShell, these commands
```

  are known as _cmdlets_ (pronounced "command-lets").

## Running OS-native commands

Any OS-native command can be run from the PowerShell command line. In most cases you run the command
exactly has you would in `bash` or `cmd.exe`. The following example shows running the `grep` command
in `bash` on Ubuntu Linux.

```bash
sdwheeler@circumflex:~$ grep sdwheeler /etc/passwd
sdwheeler:x:1000:1000:,,,:/home/sdwheeler:/bin/bash
sdwheeler@circumflex:~$ pwsh
PowerShell 7.2.6
Copyright (c) Microsoft Corporation.

https://aka.ms/powershell

> Type 'help' to get help.

Reviewer: Should this be here or in the next example block? It's referred to in the prose after this block but at first I went back to the preceding prose to see why it was being run.

Author: pwsh is run in bash. The output make more sense here.

```

After starting PowerShell on Ubuntu, you can run the same command from the PowerShell command line:

```powershell
PS /home/sdwheeler> grep sdwheeler /etc/passwd
sdwheeler:x:1000:1000:,,,:/home/sdwheeler:/bin/bash
```

> ### Passing arguments to native commands

Reviewer: ```suggestion
### Passing arguments to OS-native commands
```

Most shells include features for using variables, evaluating expressions, and handling strings. But
each shell does these differently. In PowerShell, all parameters are start with a hyphen (`-`)

> character. In `cmd.exe`, most parameter use a slash (`/`) character. Other command-line tools may

Reviewer: ```suggestion
character. In `cmd.exe`, most parameters use a slash (`/`) character. Other command-line tools may
```

not have a special character for parameters.

Each shell has its own way of handling and evaluating strings on the command line. When running
OS-native commands in PowerShell that expect strings to be quoted in a specific way, you may need
adjust how you pass those strings.

For more information, see the following articles:

- [about_Parsing][1]
- [about_Quoting_Rules][2]

PowerShell 7.2 introduced a new experimental feature `PSNativeCommandArgumentPassing` that improved

> native command handling. For more information, see [PSNativeCommandArgumentPassing][3].

Reviewer: ```suggestion
OS-native command handling. For more information, see [PSNativeCommandArgumentPassing][3].
```

### Handling output and errors

PowerShell also has several more output streams than other shells. The `bash` and `cmd.exe` shells
have **stdout** and **stderr**. PowerShell has six output streams. For more information, see
[about_Redirection][4] and [about_Output_Streams][5].

In general, the output sent to **stdout** by an OS-native command is sent to the **Success** stream
in PowerShell. Output sent to **stderr** by an OS-native command is sent to the **Error** stream in
PowerShell.

> When a native command has a non-zero exit code, `$?` is set to `$false`. If the exit code is zero,

Reviewer: ```suggestion
When an OS-native command has a non-zero exit code, `$?` is set to `$false`. If the exit code is zero,
```

`$?` is set to `$true`.

> However, this changed in PowerShell 7.2. Error records redirected from native commands, like when

Reviewer: ```suggestion
However, this changed in PowerShell 7.2. Error records redirected from OS-native commands, like when
```

using redirection operators (`2>&1`), are not written to PowerShell's `$Error` variable and the
preference variable `$ErrorActionPreference` does not affect the redirected output.

> Many native commands write to `stderr` as an alternative stream for additional information. This

Reviewer: Above we wrap stream names in `**` - should that be the case here too?
```suggestion
Many native commands write to **stderr** as an alternative stream for additional information. This
```

behavior can cause confusion in PowerShell when looking through errors and the additional output
information can be lost if `$ErrorActionPreference` is set to a state that mutes the output.

PowerShell 7.3 added a new experimental feature `PSNativeCommandErrorActionPreference` that allows
you to control whether output to `stderr` is treated as an error. For more information, see
[PSNativeCommandErrorActionPreference][6].

## Running PowerShell commands

As previous noted, PowerShell commands are are known as cmdlets. Cmdlets are collected into
PowerShell modules that can be loaded on demand. Cmdlets can be written in any compiled .NET
language or using the PowerShell scripting language itself.

### PowerShell commands that run other commands

The PowerShell **call operator** (`&`) lets you run commands that are stored in variables and

> represented by strings or script blocks. You can use this to run any native command or PowerShell

Reviewer: ```suggestion
represented by strings or script blocks. You can use this to run any OS-native command or PowerShell
```

command. This is useful in a script when you need to dynamically construct the command-line

> parameters for a native command. For more information, see the [call operator][7].

Reviewer: ```suggestion
parameters for an OS-native command. For more information, see the [call operator][7].
```

> The `Start-Process` cmdlet can be used to run native commands, but should only be used when you need

Reviewer: ```suggestion
The `Start-Process` cmdlet can be used to run OS-native commands, but should only be used when you need
```

to control how the command is executed. The cmdlet has parameters to support the following
scenarios:

- Run a command using different credentials
- Hide the console window created by the new process
- Redirect **stdin**, **stdout**, and **stderr** streams
- Use a different working directory for the command

> The following example runs the native command `sort.exe` with redirected input and output streams.

Reviewer: ```suggestion
The following example runs the OS-native command `sort.exe` with redirected input and output streams.
```

```powershell
$processOptions = @{
    FilePath = "sort.exe"
    RedirectStandardInput = "TestSort.txt"
    RedirectStandardOutput = "Sorted.txt"
    RedirectStandardError = "SortError.txt"
    UseNewEnvironment = $true
}
Start-Process @processOptions
```

For more information, see [Start-Process][8].

> On Windows, the `Invoke-Item` command performs the default action for the specified item. For

Reviewer: I think using `cmdlet` over `command` here can help disambiguate between OS-native commands and PowerShell commands a bit.
```suggestion
On Windows, the `Invoke-Item` cmdlet performs the default action for the specified item. For
```

example, it runs an executable file or opens a document file using the application associated with
the document file type. The default action depends on the type of item and is resolved by the
PowerShell provider that provides access to the item.

> The following example opens the PowerShell source code respoitory in your default web browser.

Author: ```suggestion
The following example opens the PowerShell source code repository in your default web browser.
```

```powershell

> Invoke-Item https://github.com/PowerShell/PowerShell

Reviewer: Nit: No parameter here. Is that to simplify things for the reader?

Author: Yes. The parameter would be `-Path`. That could be confusing in this context since it is not a file path.
